File: src/omnia/data/structure/tuple/Octuplet.java
package omnia.data.structure.tuple;

import java.util.function.Function;

public interface Octuplet<T> extends Octuple<T, T, T, T, T, T, T, T>, Tuples.AtMostOctuplet<T> {

  static <T> Octuplet<T> of(T first, T second, T third, T fourth, T fifth, T sixth, T seventh, T eighth) {
    return new ImmutableOctuplet<>(first, second, third, fourth, fifth, sixth, seventh, eighth);
  }

  @Override
  <R> Octuplet<R> map(Function<? super T, ? extends R> mapper);

  @Override
  Septuplet<T> dropFirst();

  @Override
  Septuplet<T> dropSecond();

  @Override
  Septuplet<T> dropThird();

  @Override
  Septuplet<T> dropFourth();

  @Override
  Septuplet<T> dropFifth();

  @Override
  Septuplet<T> dropSixth();

  @Override
  Septuplet<T> dropSeventh();

  @Override
  Septuplet<T> dropEighth();

  @Override
  Nonuplet<T> concat(T object);

  @Override
  Decuplet<T> concat(Couple<T, T> other);
}

File: src/omnia/data/structure/tuple/Octuplet.kt
package omnia.data.structure.tuple

import omnia.data.structure.tuple.Tuples.AtMostOctuplet

interface Octuplet<T : Any> : Octuple<T, T, T, T, T, T, T, T>, AtMostOctuplet<T> {

  override fun <R : Any> map(mapper: (T) -> R): Octuplet<R>
  override fun dropFirst(): Septuplet<T>
  override fun dropSecond(): Septuplet<T>
  override fun dropThird(): Septuplet<T>
  override fun dropFourth(): Septuplet<T>
  override fun dropFifth(): Septuplet<T>
  override fun dropSixth(): Septuplet<T>
  override fun dropSeventh(): Septuplet<T>
  override fun dropEighth(): Septuplet<T>
  override fun concat(`object`: T): Nonuplet<T>
  override fun concat(other: Couple<T, T>): Decuplet<T>

  companion object {

    fun <T : Any> of(
        first: T,
        second: T,
        third: T,
        fourth: T,
        fifth: T,
        sixth: T,
        seventh: T,
        eighth: T
    ): Octuplet<T> {
      return ImmutableOctuplet(first, second, third, fourth, fifth, sixth, seventh, eighth)
    }
  }
}